File: plugin_src/intent/src/ios/IntentPlugin.swift
 /*
  * Copyright (c) 2021 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

 import Foundation

 @objc(IntentPlugin)
 class IntentPlugin : CDVPlugin {
    var callbackId: String? = nil;
    var msgListener: ((Int, String, String)->(Void))? = nil;
    var intentCallbackId: String? = nil;
    var intentListener:  ((String, String?, String, Int64)->(Void))? = nil;

    var isLauncher = false;
    var isChangeIconPath = false;

    override func pluginInitialize() {
        IntentManager.getShareInstance().setViewController(viewController as! CDVViewController, self.commandDelegate);
    }

    //---------------------------------------------------------
    func success(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK)

        self.commandDelegate?.send(result, callbackId: command.callbackId)
    }

    func success(_ command: CDVInvokedUrlCommand, _ retAsString: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK,
                                     messageAs: retAsString);

        self.commandDelegate?.send(result, callbackId: command.callbackId)
    }

    func success(_ command: CDVInvokedUrlCommand, retAsDict: [String : Any]) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK,
                                     messageAs: retAsDict);

        self.commandDelegate?.send(result, callbackId: command.callbackId)
    }

    func success(_ command: CDVInvokedUrlCommand, retAsArray: [String]) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK,
                                     messageAs: retAsArray);

        self.commandDelegate?.send(result, callbackId: command.callbackId)
    }

    func error(_ command: CDVInvokedUrlCommand, _ retAsString: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR,
                                     messageAs: retAsString);

        self.commandDelegate?.send(result, callbackId: command.callbackId)
    }

    func sendCallback(_ command: CDVInvokedUrlCommand, _ status: CDVCommandStatus, _ keepCallback:Bool, _ retAsString: String?) {
        var result: CDVPluginResult? = nil;
        if (status != CDVCommandStatus_NO_RESULT) {
            result = CDVPluginResult(status: status, messageAs: retAsString);
        }
        else {
            result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
        }
        result?.setKeepCallbackAs(keepCallback);

        self.commandDelegate?.send(result, callbackId: command.callbackId)
    }


    @objc func sendIntent(_ command: CDVInvokedUrlCommand) {
        let action = command.arguments[0] as? String ?? "";
        let params = command.arguments[1] as? String ?? nil;

        let info = IntentInfo(action, params, command.callbackId);

        do {
            try IntentManager.getShareInstance().sendIntent(info);
            let result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
            result?.setKeepCallbackAs(true);
            self.commandDelegate?.send(result, callbackId: command.callbackId)
        } catch let error {
            self.error(command, error.localizedDescription);
        }
    }

    @objc func sendUrlIntent(_ command: CDVInvokedUrlCommand) {
        let url = command.arguments[0] as? String ?? "";
        let uri = URL(string: url);
        if (uri == nil) {
            self.error(command, "Url is invalid!");
            return;
        }

        do {
            if (IntentManager.getShareInstance().isInternalIntent(url)) {
                try IntentManager.getShareInstance().receiveIntent(uri!, command.callbackId);
                let result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
                result?.setKeepCallbackAs(true);
                self.commandDelegate?.send(result, callbackId: command.callbackId)
            }
            else {
                try IntentManager.getShareInstance().openUrl(url);
                self.success(command);
            }
        }
        catch let error {
            self.error(command, error.localizedDescription);
        }
    }

    @objc func sendIntentResponse(_ command: CDVInvokedUrlCommand) {
        let result = command.arguments[0] as? String ?? "";
        let intentId = command.arguments[1] as? Int64 ?? -1

        do {
            try IntentManager.getShareInstance().sendIntentResponse(result, intentId);
            self.success(command);
        } catch let error {
            self.error(command, error.localizedDescription);
        }
    }

    @objc func addIntentListener(_ command: CDVInvokedUrlCommand) {
        self.intentCallbackId = command.callbackId;
        // Don't return any result now
        let result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
        result?.setKeepCallbackAs(true);
        self.commandDelegate?.send(result, callbackId: command.callbackId)
        try? IntentManager.getShareInstance().setListenerReady(command.callbackId);
    }
 }

File: plugin_src/intent/src/android/IntentPlugin.java
/*
 * Copyright (c) 2021 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.essentials.plugins.intent;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;

public class IntentPlugin extends CordovaPlugin {
    private static final String LOG_TAG = "Intent";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        try {
            switch (action) {
                case "sendIntent":
                    this.sendIntent(args, callbackContext);
                    break;
                case "sendUrlIntent":
                    this.sendUrlIntent(args, callbackContext);
                    break;
                case "addIntentListener":
                    this.addIntentListener(callbackContext);
                    break;
                case "sendIntentResponse":
                    this.sendIntentResponse(args, callbackContext);
                    break;

                default:
                    return false;
            }
        } catch (Exception e) {
            e.printStackTrace();
            callbackContext.error(e.getLocalizedMessage());
        }
        return true;
    }

    private void getIntentUri(Intent intent) {
        if (intent == null) {
            return;
        }

        String action = intent.getAction();
        if ((action != null) && action.equals("android.intent.action.VIEW")) {
            Uri uri = intent.getData();
            if (uri != null) {
                IntentManager.getShareInstance().setIntentUri(uri);
            }
        }
    }

    @Override
    public void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        getIntentUri(intent);
    }

    @Override
    protected void pluginInitialize() {
        Activity activity = cordova.getActivity();
        IntentManager.getShareInstance().setActivity(activity, preferences);
        getIntentUri(activity.getIntent());
        super.pluginInitialize();
    }

    protected void sendIntent(JSONArray args, CallbackContext callbackContext) throws Exception {
        String action = args.getString(0);
        String params = args.getString(1);

        IntentInfo info = new IntentInfo(action, params, callbackContext);
        IntentManager.getShareInstance().sendIntent(info);
        PluginResult pluginResult = new PluginResult(PluginResult.Status.NO_RESULT);
        pluginResult.setKeepCallback(true);
        callbackContext.sendPluginResult(pluginResult);
    }

    protected void sendUrlIntent(JSONArray args, CallbackContext callbackContext) throws Exception {
        String url = args.getString(0);

        if (IntentManager.getShareInstance().isInternalIntent(url)) {
            IntentManager.getShareInstance().receiveIntent(Uri.parse(url), callbackContext);
            PluginResult pluginResult = new PluginResult(PluginResult.Status.NO_RESULT);
            pluginResult.setKeepCallback(true);
            callbackContext.sendPluginResult(pluginResult);
        }
        else {
            IntentManager.getShareInstance().showWebPage(url);
            callbackContext.success();
        }
    }

    protected void sendIntentResponse(JSONArray args, CallbackContext callbackContext) throws Exception {
        String result = args.getString(0);
        long intentId = args.getLong(1);

        IntentManager.getShareInstance().sendIntentResponse(result, intentId);
        callbackContext.success();
    }

    protected void addIntentListener(CallbackContext callbackContext) throws Exception {
        PluginResult pluginResult = new PluginResult(PluginResult.Status.NO_RESULT);
        pluginResult.setKeepCallback(true);
        callbackContext.sendPluginResult(pluginResult);
        IntentManager.getShareInstance().setListenerReady(callbackContext);
    }
}
